src/native/libs/System.Globalization.Native/pal_collation.m

```diff
@@ -59,19 +59,170 @@ int32_t GlobalizationNative_CompareStringNative(const uint16_t* localeName, int3
         currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:locName];
     }
 
-    NSString *firstString = [NSString stringWithCharacters: lpStr1 length: cwStr1Length];
-    NSString *secondString = [NSString stringWithCharacters: lpStr2 length: cwStr2Length];
-    NSRange string1Range = NSMakeRange(0, cwStr1Length);
+    NSString *sourceString = [NSString stringWithCharacters: lpSource length: cwSourceLength];
+    NSString *sourceStrComposed = sourceString.precomposedStringWithCanonicalMapping;
+    NSString *targetString = [NSString stringWithCharacters: lpTarget length: cwTargetLength];
+    NSString *targetStrComposed = targetString.precomposedStringWithCanonicalMapping;
+
+    NSRange comparisonRange = NSMakeRange(0, sourceStrComposed.length);
+    NSStringCompareOptions options = ConvertFromCompareOptionsToNSStringCompareOptions(comparisonOptions);
+    
+    // in case mapping is not found
+    if (options == 0)
+        return -2;
+
+    return [sourceStrComposed compare:targetStrComposed
+                              options:options
+                              range:comparisonRange
+                              locale:currentLocale];
+}
+
+NSString* ComposeString(NSString* str)
+{
+    NSString* source = str.precomposedStringWithCanonicalMapping;
+    // Below we are removing weightless characters from the string to get ICU behavior.
+    NSString* zarb = @"\u200d";
+    NSString* nullChar = @"\0";
+    // Remove zero width joiner
+    NSString* result = [source stringByReplacingOccurrencesOfString:zarb withString:@""];
+    // Remove null characters
+    result = [result stringByReplacingOccurrencesOfString:nullChar withString:@""];
+
+    return result;
+}
```

Comment:
Are these the only weightless characters? What is the difference between this function and `precomposedStringWithCanonicalMapping` that is used above?

Reply:
Right, what about `\u200B`, `\u200C`, `\uFEFF`? Isn't regexing once like here https://github.com/dotnet/runtime/blob/89f3a9ef41383bb409b69d1a0f0db910f3ed9a34/src/mono/wasm/runtime/hybrid-globalization/collations.ts#L282 faster than traversing separately for each char?

Reply:
Updated with `\u200B-\u200D\uFEFF\0`.